src/cmd/flash.js

```diff
@@ -17,153 +16,132 @@ const systemModuleIndexToString = {
-			return when.reject();
+			// TODO: Replace by UsageError
+			return Promise.reject();
 		}
 
 		let result;
-		if (this.options.usb) {
-			result = this.flashDfu();
-		} else if (this.options.serial) {
-			result = this.flashYModem();
+		if (usb) {
+			result = this.flashDfu({ binary, factory, force });
+		} else if (serial) {
+			result = this.flashYModem({ binary });
 		} else {
-			result = this.flashCloud();
+			result = this.flashCloud({ device, files, target, yes });
 		}
 
 		return result;
 	}
 
-	flashCloud() {
-		const args = {
-			params: {
-				device: this.options.params.device,
-				files: this.options.params.files,
-			}
-		};
+	flashCloud({ device, files, target, yes }) {
 		const CloudCommands = require('../cmd/cloud');
-		return new CloudCommands(args).flashDevice();
+		return new CloudCommands().flashDevice(device, files, { target, yes });
 	}
 
-	flashYModem() {
+	flashYModem({ binary }) {
 		const args = {
 			params: {
-				binary: this.options.params.binary
+				binary: binary
 			}
 		};
 		const SerialCommands = require('../cmd/serial');
 		return new SerialCommands(args).flashDevice();
 	}
 
-	flashDfu() {
-		const useFactory = this.options.useFactoryAddress;
-		let firmware = this.options.params.binary;
-
+	flashDfu({ binary, factory, force }) {
 		let specs, destSegment, destAddress;
 		let flashingKnownApp = false;
-		const ready = sequence([
-			() => {
-				return dfu.isDfuUtilInstalled();
-			},
-			() => {
-				return dfu.findCompatibleDFU();
-			},
-			() => {
-				//only match against knownApp if file is not found
-				let stats;
-				try {
-					stats = fs.statSync(firmware);
-				} catch (ex) {
-					// file does not exist
-					firmware = dfu.checkKnownApp(firmware);
-					if (firmware === undefined) {
-						return when.reject('file does not exist and no known app found.');
-					} else {
-						flashingKnownApp = true;
-						return firmware;
-					}
-				}
-
-				if (!stats.isFile()){
-					return when.reject('You cannot flash a directory over USB');
-				}
-			},
-			() => {
-				destSegment = useFactory ? 'factoryReset' : 'userFirmware';
-				if (flashingKnownApp) {
-					return when.resolve();
+		return Promise.resolve().then(() => {
+			return dfu.isDfuUtilInstalled();
+		}).then(() => {
+			return dfu.findCompatibleDFU();
+		}).then(() => {
+			//only match against knownApp if file is not found
+			let stats;
+			try {
+				stats = fs.statSync(binary);
+			} catch (ex) {
+				// file does not exist
+				binary = dfu.checkKnownApp(binary);
+				if (binary === undefined) {
+					throw new Error('file does not exist and no known app found.');
+				} else {
+					flashingKnownApp = true;
+					return binary;
 				}
+			}
 
-				return when.promise((resolve, reject) => {
-					const parser = new ModuleParser();
-					parser.parseFile(firmware, (info, err) => {
-						if (err) {
-							return reject(err);
-						}
+			if (!stats.isFile()){
+				throw new Error('You cannot flash a directory over USB');
```

Comment:
`new VError`? Not sure if we should replace it, are there cases where we should use Error?